File: android/src/main/java/com/damly/voice/recognizer/VoiceRecognizerJsEvent.java
package com.damly.voice.recognizer;

import android.content.Context;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import javax.annotation.Nullable;

/**
 * Created by damly on 2018/1/17.
 */

public class VoiceRecognizerJsEvent {

    private ReactContext mContext = null;

    public VoiceRecognizerJsEvent(ReactContext context) {
        mContext = context;
    }

    public void onVolumeChanged(int volume) {
        WritableMap params = Arguments.createMap();
        params.putInt("volume", volume);
        onJSEvent(mContext, "onVoiceRecognizerVolumeChanged", params);
    }

    public void onCancel() {
        onJSEvent(mContext, "onVoiceRecognizerCancel", null);
    }

    public void onResult(String result, boolean isLast, String provider) {
        WritableMap params = Arguments.createMap();
        params.putString("result", result);
        params.putBoolean("isLast", isLast);
        params.putString("provider", provider);
        onJSEvent(mContext, "onVoiceRecognizerResult", params);
    }

    public void onError(int errorCode, String description){
        WritableMap params = Arguments.createMap();

        params.putInt("errorCode", errorCode);
        params.putString("message", description);

        onJSEvent(mContext, "onVoiceRecognizerError", params);
    }

    private void onJSEvent(ReactContext reactContext,
                           String eventName,
                           @Nullable WritableMap params) {
        reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }
}

File: ios/VoiceRecognizerJsEvent.m
//
//  VoiceRecognizerJsEvent.m
//  VoiceManager
//
//  Created by damly on 2018/1/19.
//  Copyright © 2018年 damly. All rights reserved.
//

#import "VoiceRecognizerJsEvent.h"
#import <Foundation/Foundation.h>
#import <React/RCTUtils.h>
#import <React/RCTEventDispatcher.h>

@implementation VoiceRecognizerJsEvent
{
    RCTBridge *_bridge;
}

- (id) initWith:(RCTBridge *)bridge
{
    self = [super init];
    
    _bridge = bridge;
    
    return self ;
}

- (id) init
{
    return [self initWith:nil] ;
}


- (void) onResult:(NSString *)result isLast:(BOOL) isLast provider:(NSString *)provider
{
    if(_bridge != nil)
    {
        NSDictionary * event = @{
                                 @"result": result,
                                 @"isLast": [NSNumber numberWithBool: isLast],
                                 @"provider": provider,
                                 };
        [_bridge.eventDispatcher sendAppEventWithName:@"onVoiceRecognizerResult" body:event];
    }
}

- (void) onVolumeChanged:(int) volume
{
    if(_bridge != nil)
    {
        NSDictionary * event = @{
                                 @"volume": [NSNumber numberWithInt: volume]
                                 };
        [_bridge.eventDispatcher sendAppEventWithName:@"onVoiceRecognizerVolumeChanged" body:event];
    }
}

- (void) onCancel
{
    if(_bridge != nil)
    {
        [_bridge.eventDispatcher sendAppEventWithName:@"onVoiceRecognizerCancel" body:nil];
    }
}

- (void) onError:(int) code message:(NSString *) message
{
    if(_bridge != nil)
    {
        NSDictionary * event = @{
                                 @"errorCode": [NSNumber numberWithInt: code],
                                 @"errorDesc": message,
                                 };
        [_bridge.eventDispatcher sendAppEventWithName:@"onVoiceRecognizerError" body:event];
    }
}

@end
